### ISSUE TITLE
Stuck at applying iptables rules.

### ISSUE THREAD

**Issue by code-issues-pull** [ukanth/afwall#510]
Android v5.1.1 (Nexus6 so it's stock).
SuperSU v2.65 (logs show that AFWall+ requested root and it was granted)
Afwall+ v2.2.2 Donate

Upon startup, Afwall+ error (Error applying firewall rules.) message.

Entering the afwall shows it is enabled and if I have blocked something, it does appear that the rule is in effect to block it.

Disabling the firewall and reenabling it results in it hanging at the Working message (Applying iptables rules).

This problem only started in the last month or so, and I've left AFwall disabled since it showed up as I have been too busy to mess with it.

**Comment by code-issues-pull** [comment]
Well, decided to clear cache and app data for AFWall and this appears to have resolved the issue. As to what caused it in the first place, no idea.

**Comment by code-issues-pull** [comment]
Setting a custom script is hanging it. ". /sdcard/Scripts/net_settings.sh" is what I put in.

Script contains:
echo 0 > /proc/sys/net/ipv4/conf/all/accept_redirects

echo 0 > /proc/sys/net/ipv4/tcp_timestamps
echo 1 > /proc/sys/net/ipv6/conf/wlan0/disable_ipv6

setprop dhcp.tiwlan0.dns1 23.226.230.72
setprop dhcp.tiwlan0.dns2 173.230.156.28
setprop net.ppp0.dns1 23.226.230.72
setprop net.ppp0.dns2 173.230.156.28
setprop net.dns1 23.226.230.72
setprop net.dns2 173.230.156.28
setprop net.rmnet0.dns1 23.226.230.72
setprop net.rmnet0.dns2 173.230.156.28
setprop net.pdpbr1.dns1 23.226.230.72
setprop net.pdpbr1.dns2 173.230.156.28

**Comment by ukanth** [comment]
Please provide logcat while it's in hang state.

**Comment by ecanja** [comment]
_tl;dr:_

I have the same issue, but only while using some kind of custom script. Moreover, in my case it's clearly connected to issue #452, as I get the notification mentioned there after a reboot.
Anyway, the logcat excerpt from adb logcat while hanging is as follows:

```
[...]
V/WindowManager( 2690): not Base app: Adding window Window{d8ab33a u0 dev.ukanth.ufirewall/dev.ukanth.ufirewall.MainActivity} at 3 of 10
E/AFWall  ( 3713): Exception : java.util.NoSuchElementException
I/Timeline( 3713): Timeline: Activity_idle id: android.os.BinderProxy@1697190 time:574608
I/Timeline( 2690): Timeline: Activity_windows_visible id: ActivityRecord{d9600bf u0 dev.ukanth.ufirewall/.MainActivity t416} time:574839
W/ProcessCpuTracker( 2690): Skipping unknown process pid 5912
I/ThermalEngine( 1998): Sensor:pm8941_tz:37676 mC
I/AFWall  ( 3713): [libsuperuser] [SU%] WATCHDOG_EXIT
[...]
```

_More detailed report:_

I have two custom scripts, a long and a short one. Both contain a preamble (including `IPTABLES=/system/bin/iptables` and some flushing rules) and some short rules for Orbot, DNS and NTP (please let me know, if you need these). That's it for the short one. The long script additionally contains a huge amount (~800 lines) of rules looking like this:

`$IPTABLES -A "afwall" -d <IP>/>Port> -j REJECT`

and some lines in the form of

`$IPTABLES -A "afwall" -m iprange --dst-range <IP1>-<IP2>`

After enabling the firewall and applying this long script, I get stuck and the app doesn't react anymore. After tapping the home button and then starting AFWall again, the app is still stuck at the popup. Apart from killing AFWall, the following procedure strangely worked as well: Before applying the script, I enabled the notification icon in the AFWall settings. After getting stuck and tapping the home button, I tap on the AFWall notification and a new AFWall window opens. This is apparently a second window (where e.g. getting logs or rules via the menu doesn't work - they're just blank), because after tapping the back button, I'm redirected to the first window, where now the popup has disappeared. By the way, I can even iterate this behaviour with a third window! And as a side note: After tapping the notification, `adb logcat` provided ~400 lines less output than before tapping it. And it wasn't just one big block missing (I haven't investigated this further).

After this procedure, AFWall seems to have restarted - the log looks as follows:

```
======
Logcat
======

14:51:25 Starting /data/data/dev.ukanth.ufirewall/app_bin/klogripper 
14:51:25 Starting root shell...
14:51:25 [libsuperuser] [SU%] START
14:51:25 Root shell is open
14:51:25 logservice shell started
```

And to my surprise, all rules of my script are listed in the "Firewall Rules", so they even seem to have been applied. After a reboot, I got:

```
======
Logcat
======

14:56:21 Now assuming NO connection (all interfaces down)
14:56:21 Exception : java.util.NoSuchElementException
14:56:21 Starting root shell...
14:56:21 [libsuperuser] [SU%] START
14:56:21 Starting /data/data/dev.ukanth.ufirewall/app_bin/klogripper 
14:56:21 [libsuperuser] [SU%] START
14:56:22 Root shell is open
14:56:22 logservice shell started
14:56:28 [libsuperuser] [SU%] WATCHDOG_EXIT
14:56:28 Exception : java.util.NoSuchElementException
14:56:28 libsuperuser error -1 on command '. /data/afwall_scripts/iptables.sh'
14:56:28 Exception : java.util.NoSuchElementException
15:01:47 Starting root shell...
15:01:47 [libsuperuser] [SU%] START
15:01:47 Root shell is open
```

(By the way, the `Exception : java.util.NoSuchElementException` message also appears when just tapping the "apply" or "enable firewall" buttons independent of any script choices.)

I guess, the interesting lines are

```
14:56:28 [libsuperuser] [SU%] WATCHDOG_EXIT
14:56:28 libsuperuser error -1 on command '. /data/afwall_scripts/iptables.sh'
```

Does this help somehow?

To get a logcat while it's hanging, I needed to use `adb logcat` (as the AFWall logs got cleared) - once before applying the script, once afterwards. The comparison of the two files gives the output difference posted in the beginning.

NOTE: If I remove some of the REJECT rules (but not all of them - I left ~200 lines), the issue disappears. So, 800 lines may just be too much. Or they contain a syntax error, in which case it's not AFWall's failure, but mine and my report might be superfluous. ;-) I nevertheless post it, as it may help to debug the issue.

EDIT: Today, the problem also occurred with the ~200 lines script. So after all, it does seem random ...

**Comment by ghost** [comment]
This happens to me too (with latest version afwall), after phone reboot! only after random reboots, and im not using any script.
it gets stuck and then i have to disable and re-enable firewall rules, otherwise my internet doesnt work.

**Comment by T-vK** [comment]
It happens to me too. For me it is related to WiFi/Data. When either WiFI or Data was enabled before rebooting then AFWall+ shows an error notification "Error applying firewall rules" and when I manually tap on Apply Rules, then it just hangs forever in the "Applying iptable rules"-screen. To manually fix this, I close AFWall+ and then disable WiFi and Data and then open AFWall+ again and Apply Rules, then enable WiFi/Data again. If I turn WiFi and Data off before rebooting, then AFWall+ works just fine. But it's annoying because you can easily forget and you have to manually enable WiFi/Data...
